### ISSUE TITLE
New provider powerdns (pdns)

### ISSUE THREAD

**Issue by samip5** [cloudydeno/kubernetes-dns-sync#2]
It would be awesome if you could add pdns provider too. :)

**Comment by danopia** [comment]
Hi, thanks for the interest and I hope you see something here that draws interest because otherwise this project is a bit bare bones :) as you are the first person to file anything on it yet.

It looks like this is the API of interest, specifically to update records? https://doc.powerdns.com/authoritative/http-api/zone.html#patch--servers-server_id-zones-zone_id

Providers should be pretty simple integrations however I'd still want to verify on my own before uploading code so fortunately it looks like this is a process that I can run locally for testing.

**Comment by samip5** [comment]
That would be correct after finding out the external-dns short comings regarding IPv6 support.

**Comment by danopia** [comment]
I just pushed what appears to be working PowerDNS support for **A/AAAA/TXT**, in case you're still interested in trying out this project. The latest container at `ghcr.io/cloudydeno/kubernetes-dns-sync` includes this support.

I tested with this `config.toml` file:

```toml
[[source]]
type = "crd"

[[source]]
type = "ingress"

[[provider]]
type = "powerdns"
api_endpoint = "http://localhost:7070/api/"

[registry]
type = "txt"
txt_prefix = "_registry."
txt_owner_id = "demo"
```

The API key is read from the `POWERDNS_API_KEY` environment variable.

The zones have to exist already because I'm only managing rrsets. 

`MX` and `CNAME` records **do not work** in this first pass (though they'll be pretty easy to fix, just formatting issues).

**Comment by samip5** [comment]
Do you have an Helm chart for this yet?

**Comment by samip5** [comment]
Do you have an Helm chart for this yet?

**Comment by samip5** [comment]
Do you have an Helm chart for this yet?

**Comment by danopia** [comment]
No, sorry, I don't use helm on my personal cluster, rather `kustomize`.

* The easiest way to test stuff quickly is probably installing [Deno](https://deno.land/) and running a command like `POWERDNS_API_KEY=changeme deno run --allow-all controller/mod.ts --once` so the program runs once interactively and exits. 

* If you want to test on-cluster, first run with `--dry-run` instead so the program loops & prints but doesn't try making any actual changes. (The default is to prompt for changes, which isn't useful in a pod.)

    I should document running on-cluster but it hasn't come up yet, so below are some YAML snippets.

If you have further questions about running this, we could 'huddle' on the Kubernetes slack.

The container would be something like:

```yaml
      - name: app
        image: ghcr.io/cloudydeno/kubernetes-dns-sync
        args:
        - --dry-run
        envFrom:
        - secretRef:
            name: dns-sync-credentials-d49ch6h2f7
        securityContext:
          capabilities:
            drop:
            - ALL
        volumeMounts:
        - mountPath: /src/kubernetes-dns-sync/config.toml
          name: cfgmap
          subPath: config.toml
```

The RBAC from external-dns is probably fine, but for completeness here's my `ClusterRole`, this includes grants for the cert-manager ACME integration.

```yaml
rules:
- apiGroups:
  - ""
  resources:
  - services
  - pods
  - nodes
  verbs:
  - get
  - watch
  - list
- apiGroups:
  - networking.k8s.io
  resources:
  - ingresses
  verbs:
  - get
  - watch
  - list
- apiGroups:
  - externaldns.k8s.io
  resources:
  - dnsendpoints
  verbs:
  - get
  - watch
  - list
- apiGroups:
  - externaldns.k8s.io
  resources:
  - dnsendpoints/status
  verbs:
  - '*'
- apiGroups:
  - acme.cert-manager.io
  resources:
  - challenges
  verbs:
  - get
  - watch
  - list
- apiGroups:
  - acme.cert-manager.io
  resources:
  - challenges/status
  verbs:
  - '*'
```

**Comment by samip5** [comment]
> No, sorry, I don't use helm on my personal cluster, rather `kustomize`.
> 
> * The easiest way to test stuff quickly is probably installing [Deno](https://deno.land/) and running a command like `POWERDNS_API_KEY=changeme deno run --allow-all controller/mod.ts --once` so the program runs once interactively and exits.
> * If you want to test on-cluster, first run with `--dry-run` instead so the program loops & prints but doesn't try making any actual changes. (The default is to prompt for changes, which isn't useful in a pod.)
>   I should document running on-cluster but it hasn't come up yet, so below are some YAML snippets.
> 
> If you have further questions about running this, we could 'huddle' on the Kubernetes slack.
> 
> The container would be something like:
> 
> ```yaml
>       - name: app
>         image: ghcr.io/cloudydeno/kubernetes-dns-sync
>         args:
>         - --dry-run
>         envFrom:
>         - secretRef:
>             name: dns-sync-credentials-d49ch6h2f7
>         securityContext:
>           capabilities:
>             drop:
>             - ALL
>         volumeMounts:
>         - mountPath: /src/kubernetes-dns-sync/config.toml
>           name: cfgmap
>           subPath: config.toml
> ```
> 
> The RBAC from external-dns is probably fine, but for completeness here's my `ClusterRole`, this includes grants for the cert-manager ACME integration.
> 
> ```yaml
> rules:
> - apiGroups:
>   - ""
>   resources:
>   - services
>   - pods
>   - nodes
>   verbs:
>   - get
>   - watch
>   - list
> - apiGroups:
>   - networking.k8s.io
>   resources:
>   - ingresses
>   verbs:
>   - get
>   - watch
>   - list
> - apiGroups:
>   - externaldns.k8s.io
>   resources:
>   - dnsendpoints
>   verbs:
>   - get
>   - watch
>   - list
> - apiGroups:
>   - externaldns.k8s.io
>   resources:
>   - dnsendpoints/status
>   verbs:
>   - '*'
> - apiGroups:
>   - acme.cert-manager.io
>   resources:
>   - challenges
>   verbs:
>   - get
>   - watch
>   - list
> - apiGroups:
>   - acme.cert-manager.io
>   resources:
>   - challenges/status
>   verbs:
>   - '*'
> ```

So what does your actual kustomization look like for it?

**Comment by samip5** [comment]
Okey, I tested it out, doesn't work properly.

Pod yaml:
```
---
apiVersion: v1
kind: Pod
metadata:
  name: dns-sync
  namespace: networking
spec:
  serviceAccountName: sa-dns-sync
  volumes:
    - name: configmap-volume
      configMap:
        name: dns-sync-creds
  containers:
    - name: app
      image: ghcr.io/cloudydeno/kubernetes-dns-sync
      args:
        - --dry-run
      env:
        - name: POWERDNS_API_KEY
          value: ${PDNS_SAMIPSOLUTIONS_API_KEY}
      securityContext:
        capabilities:
          drop:
            - ALL
      volumeMounts:
        - mountPath: /src/kubernetes-dns-sync/config.toml
          name: configmap-volume
          subPath: config.toml
  affinity:
    nodeAffinity:
      requiredDuringSchedulingIgnoredDuringExecution:
        nodeSelectorTerms:
          - matchExpressions:
              - key: kubernetes.io/arch
                operator: In
                values:
                  - "amd64"
```
rbac.yaml:
```
---
apiVersion: v1
kind: ServiceAccount
metadata:
  name: sa-dns-sync
  namespace: networking
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: dns-sync
rules:
  - apiGroups:
      - ""
    resources:
      - services
      - pods
      - nodes
    verbs:
      - get
      - watch
      - list
  - apiGroups:
      - networking.k8s.io
    resources:
      - ingresses
    verbs:
      - get
      - watch
      - list
  - apiGroups:
      - externaldns.k8s.io
    resources:
      - dnsendpoints
    verbs:
      - get
      - watch
      - list
  - apiGroups:
      - externaldns.k8s.io
    resources:
      - dnsendpoints/status
    verbs:
      - '*'
  - apiGroups:
      - acme.cert-manager.io
    resources:
      - challenges
    verbs:
      - get
      - watch
      - list
  - apiGroups:
      - acme.cert-manager.io
    resources:
      - challenges/status
    verbs:
      - '*'
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: dns-sync
roleRef:
  kind: ClusterRole
  name: dns-sync
  apiGroup: rbac.authorization.k8s.io
subjects:
  - kind: ServiceAccount
    name: sa-dns-sync
    namespace: networking
```



Error log:
```
--- 2021-09-30T01:25:47.301Z via schedule
--> Loading desired records from 2 sources...
error: Uncaught (in promise) SyntaxError: Unexpected token p in JSON at position 4
    const resp = await this.#client.performRequest({
              ^
    at parse (<anonymous>)
    at packageData (deno:ext/fetch/22_body.js:313:16)
    at Response.json (deno:ext/fetch/22_body.js:253:18)
    at async ExternaldnsV1alpha1Api.getDNSEndpointListForAllNamespaces (https://deno.land/x/kubernetes_apis@v0.3.1/external-dns/externaldns.k8s.io@v1alpha1/mod.ts:25:15)
    at async WatchLister.listFromUpstream (file:///src/kubernetes-dns-sync/common/watch-lister.ts:20:17)
    at async WatchLister.getFreshList (file:///src/kubernetes-dns-sync/common/watch-lister.ts:31:15)
    at async CrdSource.Endpoints (file:///src/kubernetes-dns-sync/sources/crd.ts:23:22)
    at async file:///src/kubernetes-dns-sync/controller/mod.ts:43:20
    at async Promise.all (index 0)
    at async file:///src/kubernetes-dns-sync/controller/mod.ts:42:22
```

**Comment by samip5** [comment]
Yeah, had to get rid of the crd source in the example config.

**Comment by samip5** [comment]
Yeah, had to get rid of the crd source in the example config.

**Comment by samip5** [comment]
@danopia It seems you have a quite MAJOR bug in that code.

```
--> Found 45 existing records from powerdns
    Planner changes: 15 creates, 0 updates, 0 deletes
    Encoding changed ownership labels...
    <snip>
--> - Create: A grafana.skylab.fi [ "2001:67c:1104:fc00::4443" ]
--> - Create: A grafana.skylab.fi [ "185.132.53.251" ]
<snip>
--> - Create: A id.skylab.fi [ "185.132.53.250" ]
<snip>
--> - Create: TXT _registry.grafana.skylab.fi
        new: heritage=external-dns,external-dns/owner=demo,record-type/A=managed,external-dns/resource=ingress/monitoring/grafana
--> - Create: TXT _registry.id.skylab.fi
        new: heritage=external-dns,external-dns/owner=demo,record-type/A=managed,external-dns/resource=ingress/security/authentik
==> Doing no changes due to --dry-run
==> Submitting 23 creates, 0 updates, 0 deletes to powerdns ...
powerdns: Have 23 changes for skylab.fi.
    GET powerdns /api/v1/servers/localhost/zones/53.132.185.in-addr.arpa 200
    PATCH powerdns /api/v1/servers/localhost/zones/skylab.fi. 422
error: Uncaught (in promise) Error: PowerDNS HTTP 422 Unprocessable Entity: {"error": "Record grafana.skylab.fi./A '2001:67c:1104:fc00::4443': Parsing record content (try 'pdnsutil check-zone'): unable to parse IP address"}
      throw new Error(`PowerDNS HTTP ${resp.status} ${resp.statusText}: ${text}`);
      ```

**Comment by danopia** [comment]
Wow, you're right, that's so embarrassing! I've only been using AAAAs from my CRD source. I'll push a fix within a day.

some background: External-dns isn't the only software that screwed up when I tried adding an IPv6 to a node. virtual-kubelet also failed by generating corrupted patches for my dualstack node, and that isn't fixed yet either so I haven't tried IPv6 nodes again.

**Comment by danopia** [comment]
Turns out that A/AAAA detection was implemented, there was just a dumb bug. I've pushed the one-line fix.

> P.S I don't think those errors should be uncaught...

I've found it useful to have fatal errors (such as the DNS update failing) introduce a Kubernetes restart, because it's a universal signal that something is really wrong. Perhaps a flag to `--ignore-sync-errors` would be agreeable?

**Comment by samip5** [comment]
I have an request, could this use PowerDNS-Admin API?
I would preferably like an way that is not using the PowerDNS API directly, so that I could control what it has access to.

How are you using the crd? I don't use the external-dns crd so that's probably why it errored.

**Comment by samip5** [comment]
> Turns out that A/AAAA detection was implemented, there was just a dumb bug if the IPv4 came after IPv6. I've pushed the one-line fix.
> 
You hadn't accounted for the fact that it can be the first or 2nd entry? That's little silly bug to have.

> > P.S I don't think those errors should be uncaught...
> 
> I've found it useful to have fatal errors (such as the DNS update failing) introduce a Kubernetes restart, because it's a universal signal that something is really wrong. Perhaps a flag to `--ignore-sync-errors` would be agreeable?

It doesn't need a flag, but I was just wondering if it should be caught or not.

**Comment by danopia** [comment]
> You hadn't accounted for the fact that it can be the first or 2nd entry? That's little silly bug to have.

It was just a dumb copy/paste bug. https://github.com/cloudydeno/kubernetes-dns-sync/commit/b2895defa30842c5844567942779ff4efe2d0e3f

---

> I have an request, could this use PowerDNS-Admin API (https://github.com/ngoduykhanh/PowerDNS-Admin/blob/master/docs/API.md)? I would preferably like an way that is not using the PowerDNS API directly, so that I could control what it has access to.

Judging from that document the actual rrset management API is the exact same:

> We can use apikey for all calls specified in our API specification (it tries to follow powerdns API 1:1, only tsigkeys endpoints are not yet implemented), don't forget to specify Content-Type!

So I'd suggest just trying it and seeing if it works. It should be pretty straightforward to fix Provider-specific things anyway.

I'm still not sure how trailing dots work in PowerDNS (I've been using trailing dots in zone names but the 'admin api' document doesn't), is either style allowed??

---

> How are you using the crd? I don't use the external-dns crd so that's probably why it errored.

I use the CRD in my personal cluster to manage a few things:
* CDN A/AAAA records (my main website is on Firebase Hosting)
* MX and TXT records for email (my actual mail, also a subdomain or two that points to Amazon SES)
* several A/AAAA records for specific non-Kubernetes servers

I basically use the CRD to make Kubernetes the source of truth for every DNS record on every domain name I own. It's entirely optional of course. I find the flexibility quite useful for myself.

**Comment by samip5** [comment]
It was able to create the records, but unable to delete?

```
--- 2021-10-01T12:40:53.817Z via schedule
--> Loading desired records from 1 sources...
INFO: Reflector for Ingress listed upstream
    Discovered 16 desired records from ingress
--> Discovered 16 desired records overall
    GET powerdns /api/v1/servers/localhost/zones 200
    Loading existing records from powerdns ...
    GET powerdns /api/v1/servers/localhost/zones/f.0.4.0.1.1.c.7.6.0.1.0.0.2.ip6.arpa 200
    GET powerdns /api/v1/servers/localhost/zones/1.0.2.2.0.f.8.e.0.a.2.ip6.arpa 200
    GET powerdns /api/v1/servers/localhost/zones/1.c.1.c.2.2.7.0.a.2.ip6.arpa 200
    GET powerdns /api/v1/servers/localhost/zones/home.samip.fi 200
    GET powerdns /api/v1/servers/localhost/zones/128-135.193.218.185.in-addr.arpa 200
    GET powerdns /api/v1/servers/localhost/zones/4.0.1.1.c.7.6.0.1.0.0.2.ip6.arpa 200
    GET powerdns /api/v1/servers/localhost/zones/53.132.185.in-addr.arpa 200
    GET powerdns /api/v1/servers/localhost/zones/skylab.fi 200
    Recognizing ownership labels on 68 records...
--> Found 60 existing records from powerdns
    Planner changes: 0 creates, 0 updates, 1 deletes
    Encoding changed ownership labels...
--> - Delete: A id.skylab.fi [ "185.132.53.250" ]
--> - Delete: TXT _registry.id.skylab.fi
        old: heritage=external-dns,external-dns/owner=demo,record-type/A=managed,external-dns/resource=ingress/security/authentik
error: Uncaught (in promise) Error: User declined to perform provider edits
      if (prompt(`    Proceed with editing provider records?`, 'yes') !== 'yes') throw new Error(
                                                                                       ^
    at file:///src/kubernetes-dns-sync/controller/mod.ts:110:88
```

**Comment by danopia** [comment]
```
error: Uncaught (in promise) Error: User declined to perform provider edits
```

This means that it had changes, and prompted about committing them, but didn't receive "yes" on stdin. If you are running this in-cluster and would like the changes to be automatically applied then add `--yes` to the arguments. (This sort of thing would be documented if I had documentation.)

---

Also, if you are deployed then I'd recommend adding `enable_watching = true` at the top of `config.toml` so that the loop reruns immediately when ingresses are changed. This replicates `--events` in external-dns except it's more completely implemented.

You can view a version of my actual configuration [here](https://github.com/stardustapp/cloud-infra/blob/56ce1065d092f2ccececcf47399b05752f3a6e30/namespaces/ingress-internet/dns-sync/config.toml) - Just note that the `annotation_filter` syntax in that file is outdated.

**Comment by samip5** [comment]
I would like to deploy this as an actual deployment rather than a pod, could you maybe add the example manifests to the repo?

**Comment by samip5** [comment]
> I put together an example helm chart here https://github.com/cloudydeno/kubernetes-dns-sync/tree/main/examples/helm-chart and tested it with these additional values:
> 
> ```yaml
> dnsSync:
>   config: |
>     enable_watching = true
> 
>     [[source]]
>     type = "ingress"
> 
>     [[provider]]
>     type = "vultr"
> 
>     [registry]
>     type = "txt"
>     txt_prefix = "_registry."
>     txt_owner_id = "demo"
> 
>   envFrom:
>   - secretRef:
>       # this secret already exists with the api-key in it
>       name: my-dns-sync-credentials
> ```
> 
> If you have any feedback about this example chart, please open a PR or at least fresh issue about the chart.

I'm sorry but I'm not sure how I could potentially import that chart into GitOps workflow cleanly as usually done in a way like this: https://github.com/k8s-at-home/charts.

**Comment by danopia** [comment]
Today I merged #11 which includes a more complete PowerDNS provider. The rewrite is not yet tagged `latest` (and I plan to start semver soon) but if better PowerDNS support is needed, try an image such as `c790f0de0f256d5787da5c3a785aa598b57cee08` or `c790f0de0f256d5787da5c3a785aa598b57cee08-multiarch`.
